crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::rpc_module::RpcModule;
+use crate::Reader;
+use crate::{filter, query};
+use jsonrpsee::{core::RpcResult, proc_macros::rpc};
+use move_core_types::language_storage::TypeTag;
+use sui_indexer_alt_schema::objects::{StoredCoinBalanceBucket, StoredObject};
+use sui_json_rpc_types::{Coin, CoinPage};
+use sui_open_rpc::Module;
+use sui_open_rpc_macros::open_rpc;
+use sui_types::{
+    base_types::{ObjectID, SuiAddress},
+    gas_coin::GAS,
+    object::Object,
+};
+
+#[open_rpc(namespace = "suix", tag = "Coin API")]
+#[rpc(server, namespace = "suix")]
+trait CoinApi {
+    #[method(name = "getCoins")]
+    async fn get_coins(
+        &self,
+        owner: SuiAddress,
+        coin_type: Option<String>,
+        cursor: Option<ObjectID>,
+        limit: Option<usize>,
+    ) -> RpcResult<CoinPage>;
+
+    /// Return all Coin objects owned by an address.
+    #[method(name = "getAllCoins")]
+    async fn get_all_coins(
+        &self,
+        /// the owner's Sui address
+        owner: SuiAddress,
+        /// optional paging cursor
+        cursor: Option<ObjectID>,
+        /// maximum number of items per page
+        limit: Option<usize>,
+    ) -> RpcResult<CoinPage>;
+}
+
+pub(crate) struct CoinServer(pub Reader);
+
+// Need this wrapper to avoid having to implement TryInto<Coin> in indexer alt schema crate.
+struct StoredObjectWrapper(StoredObject);
+
+#[async_trait::async_trait]
+impl CoinApiServer for CoinServer {
+    async fn get_coins(
+        &self,
+        owner: SuiAddress,
+        coin_type: Option<String>,
+        cursor: Option<ObjectID>,
+        limit: Option<usize>,
+    ) -> RpcResult<CoinPage> {
+        let coin_struct_tag = if let Some(coin_type) = coin_type {
+            sui_types::parse_sui_type_tag(&coin_type).map_err(|e| {
+                jsonrpsee::types::ErrorObject::owned(
+                    1,
+                    format!("Failed to parse coin type: {}", e),
+                    None::<()>,
+                )
+            })?
+        } else {
+            GAS::type_tag()
+        };
+        self.get_coins_impl(owner, Some(coin_struct_tag), cursor, limit)
+            .await
+    }
+
+    async fn get_all_coins(
+        &self,
+        owner: SuiAddress,
+        cursor: Option<ObjectID>,
+        limit: Option<usize>,
+    ) -> RpcResult<CoinPage> {
+        self.get_coins_impl(owner, None, cursor, limit).await
+    }
+}
+
+impl CoinServer {
+    async fn get_coins_impl(
+        &self,
+        owner: SuiAddress,
+        coin_type_tag: Option<TypeTag>,
+        cursor: Option<ObjectID>,
+        limit: Option<usize>,
+    ) -> RpcResult<CoinPage> {
+        let mut conn = self.0.connect().await?;
+        let limit = self.0.cap_page_limit(limit);
+
+        // First load the coins' object ids.
+        // TODO: optimize this query according to the indexes in coin_balance_buckets table.
+        let mut query = query!("
```

Comment:
I think you would have to rewrite the entire function with a full raw string as the query.